src/placement/mapless_jump_map.c

```diff
@@ -0,0 +1,862 @@
+ *                              previously used that will be cleared when all
+ *                              the domains have been used. Important for the
+ *                              case that we must place more shards than
+ *                              domains in a particular fault domain.
+ *                              location is valid.
+ * \param[in]   shard_num       the current shard number. This is used when
+ *                              selecting a target to determine if repeated
+ *                              targets are allowed in the case that there
+ *                              are more shards than targets
+ */
+
+static void
+get_rebuild_target(struct pool_domain *root, uint32_t *target_id,
+		   struct coll_map *co_map, uint32_t *top_level_id,
+		   uint64_t key, uint8_t *dom_used, uint32_t *dom_count,
+		   int shard_num)
+{
+	int                     i;
+	uint8_t                 *coll_dom_start;
+	uint8_t                 *used_tgts = NULL;
+	uint32_t                selected_dom;
+	uint32_t                fail_num = 0;
+	uint32_t                try = 0;
+	struct pool_target      *target;
+	struct pool_domain      *target_selection;
+
+	/* Get the top level start location for bookkeeping */
+	coll_dom_start = &dom_used[co_map->coll_offset];
+	while (1) {
+
+		uint8_t skip = 0;
+		uint32_t curr_id;
+		uint32_t num_doms;
+
+		num_doms = root->do_child_nr;
+
+		/*
+		 * If all domains have been tried and failed, clear the
+		 * collision bookkeeping then choose domain  of original
+		 * failure.
+		 */
+		if (dom_count[co_map->cnt_used_offset] + 1 >= num_doms) {
+			dom_count[co_map->cnt_used_offset] = 0;
+			top_level_id = NULL;
+			for (i = 0; i < num_doms; ++i)
+				coll_dom_start[i] = 0;
+		}
+
+		/*
+		 * Choose domains using jump consistent hash until we find a
+		 * suitable domains that has not already been used.
+		 */
+		do {
+			key = crc(key, fail_num++);
+			selected_dom = jump_consistent_hash(key, num_doms);
+			target_selection = &(root->do_children[selected_dom]);
+			curr_id = target_selection->do_comp.co_id;
+		} while (top_level_id != NULL && curr_id  == *top_level_id);
+
+		/* mark this domain as used */
+		coll_dom_start[selected_dom] = 1;
+		dom_count[co_map->cnt_used_offset]++;
+
+		/* To find rebuild target we examine all targets */
+		num_doms = target_selection->do_target_nr;
+
+		D_ALLOC_ARRAY(used_tgts, num_doms);
+
+		/*
+		 * Attempt to choose a fallback target from all targets found
+		 * in this top level domain.
+		 */
+		do {
+			key = crc(key, try++);
+
+			selected_dom = jump_consistent_hash(key, num_doms);
+			target = &target_selection->do_targets[selected_dom];
+
+			*target_id = target->ta_comp.co_id;
+			/*
+			 * keep track of what targets have been tried
+			 * in case all targets in domain have failed
+			 */
+			if (used_tgts[selected_dom] == 0) {
+				skip++;
+				used_tgts[selected_dom] = 1;
+			}
+		} while (pool_target_unavail(target) && skip < num_doms);
```

Comment:
Hmm, you need consider the fseq when choosing the replacement target. See the while loop in ring_obj_remap_shards(),  I think ring and mapless can share that loop, only choosing next spare are different for these two methods()?